### DISCUSSION TITLE
I just installed starship. Why does it slow down the bash execution time?

### DISCUSSION THREAD

**Discussion by yunielrc** [starship/starship#1634] [Q&A] [answered]
starship enabled:  
```sh
❯ time for i in $(seq 1 1000000); do test 1 = 1 ; done

real	0m28.268s
user	0m26.915s
sys	0m1.365s

~ took 28s 
```
starship disabled:  
```sh
$ time for i in $(seq 1 1000000); do test 1 = 1 ; done

real	0m1.548s
user	0m1.538s
sys	0m0.022s
```

In the following video you can see the problem:  
https://youtu.be/vPT9Sxi2ZEc

**Comment by chipbuster** [comment, marked as answer]
@yunielrc Interesting catch. I wasn't aware of this before, but I don't see how to avoid it with the way starship currently operates (aside from disabling some of the shell integrations).

This appears to happen because the starship init scripts hook the DEBUG trap to provide timing information for the prompt. This trap fires at the end of every command, which is causing the slowdown you see.

Unfortunately, I don't know of a good way to fix this for the moment, because the DEBUG trap is the only mechanism we have to determine execution time for a command automatically (unlike, e.g. zsh, which has both the `precmd` and `preexec` hooks).

I did the following test: I first ran your example in a shell with only the default bashrc, then again where the DEBUG hook was trapped against the builtin `true` command. The second one took much longer:

```
chipbuster@onyx:~$ bash -c 'time for i in $(seq 1 1000000); do test 1 = 1 ; done' > /dev/null
real    0m2.235s
user    0m2.200s
sys     0m0.047s
chipbuster@onyx:~$ bash -c 'function preexec_cmd(){ true; }; trap preexec_cmd DEBUG; time for i in $(seq 1 1000000); do test 1 = 1 ; done' > /dev/null

real    0m21.201s
user    0m19.721s
sys     0m1.490s
```

**Reply by yunielrc** [reply]
Hi @chipbuster thanks for your help

Problem solved.

The expected solution:
![Screenshot from 2020-09-01 15-04-23](https://user-images.githubusercontent.com/28462092/91894984-7f9b7280-ec64-11ea-9d83-725c28da53b2.png)

The solution: 
![Screenshot from 2020-09-01 15-02-05](https://user-images.githubusercontent.com/28462092/91895093-a9ed3000-ec64-11ea-8462-9cacba370ae0.png)

I think that the community should be aware of this problem and make the decision to use the 'cmd_duration' module at the expense of sacrificing performance.

I also think that disabling the 'cmd_duration' module in the configuration should remove the 'starship_preexec' handler from the DEBUG signal.

Best Regards,
Yuniel

@matchai 
@andytom
@davidkna
@stefb69
@heyrict
@cappyzawa

**Reply by chipbuster** [reply]
> at the expense of sacrificing performance.

Do you have an example (non-synthetic) workload where this presents a real performance problem?

I ask because the example you gave is only slowed by so much because your main bash loop is executing a completely trivial built-in. Even then, the extra time amounts to 30 microseconds per loop, almost imperceptible unless you stack a lot of those commands together.

If we change the main body to a completely trivial external program (`/bin/true`), I need about 5.5 seconds to do 10,000 iterations. With starship, this time goes up to 6.5 seconds.

In other words, this doesn't appear to significantly slow things down unless you're executing massive numbers of very short commands very quickly, and this doesn't look like most shell usage that I'm aware of.

**Reply by yunielrc** [reply]
You are right, thanks
